File: packages/nodes/src/registry.ts
/**
 * The canonical catalog of node schemas. Both runtimes and the UI import
 * this registry. New nodes land here first.
 */

import type { NodeSchema } from './schema.js';

export const NODE_SCHEMAS: readonly NodeSchema[] = [
  // ──────────────────────────── DATA ────────────────────────────
  {
    id: 'data.webcam',
    name: 'Webcam',
    category: 'data',
    description: 'Live camera feed. Captures frames on demand or continuously.',
    inputs: [],
    outputs: [{ name: 'frames', type: 'image', array: true }],
    params: [
      { name: 'captureMode', kind: 'enum', options: ['manual', 'continuous'], default: 'manual' },
      { name: 'width', kind: 'number', default: 224, min: 64, max: 1024 },
      { name: 'height', kind: 'number', default: 224, min: 64, max: 1024 },
    ],
    capability: { web: true },
    executionMode: 'reactive',
    icon: 'Camera',
  },
  {
    id: 'data.upload',
    name: 'Upload Images',
    category: 'data',
    description: 'User-provided image files (drag-drop or file picker).',
    inputs: [],
    outputs: [
      { name: 'images', type: 'image', array: true },
      { name: 'labels', type: 'labels', array: true },
    ],
    params: [
      { name: 'classes', kind: 'json', default: ['class_a', 'class_b'] },
    ],
    capability: { web: true },
    executionMode: 'triggered',
    icon: 'Upload',
  },
  {
    id: 'data.mic',
    name: 'Microphone',
    category: 'data',
    description: 'Live microphone audio. Captures clips on demand.',
    inputs: [],
    outputs: [{ name: 'clips', type: 'audio', array: true }],
    params: [
      { name: 'sampleRate', kind: 'number', default: 16000 },
      { name: 'durationMs', kind: 'number', default: 2000, min: 250, max: 10000 },
    ],
    capability: { web: true },
    executionMode: 'reactive',
    icon: 'Mic',
  },

  // ──────────────────────────── BACKBONE ────────────────────────────
  {
    id: 'backbone.clip.vit-b-32',
    name: 'CLIP ViT-B/32 (image)',
    category: 'backbone',
    description: 'OpenAI CLIP ViT-B/32 image encoder. 512-d features.',
    inputs: [{ name: 'images', type: 'image', array: true }],
    outputs: [{ name: 'features', type: 'features', array: true }],
    params: [{ name: 'normalize', kind: 'boolean', default: true }],
    capability: { web: true, downloadMb: 335 },
    executionMode: 'reactive',
    icon: 'Brain',
  },
  {
    id: 'backbone.clip.text-vit-b-32',
    name: 'CLIP ViT-B/32 (text)',
    category: 'backbone',
    description: 'CLIP text encoder. 512-d features aligned with CLIP image.',
    inputs: [{ name: 'text', type: 'text', array: true }],
    outputs: [{ name: 'features', type: 'features', array: true }],
    params: [{ name: 'normalize', kind: 'boolean', default: true }],
    capability: { web: true, downloadMb: 246 },
    executionMode: 'reactive',
    icon: 'Type',
  },
  {
    id: 'backbone.dinov2.small',
    name: 'DINOv2 ViT-S/14',
    category: 'backbone',
    description: 'Meta DINOv2 self-supervised image features. 384-d.',
    inputs: [{ name: 'images', type: 'image', array: true }],
    outputs: [{ name: 'features', type: 'features', array: true }],
    params: [],
    capability: { web: true, downloadMb: 88 },
    executionMode: 'reactive',
    icon: 'Eye',
  },
  {
    id: 'backbone.mobilenet.v3',
    name: 'MobileNetV3',
    category: 'backbone',
    description: 'Lightweight mobile-first image backbone. 1280-d features.',
    inputs: [{ name: 'images', type: 'image', array: true }],
    outputs: [{ name: 'features', type: 'features', array: true }],
    params: [],
    capability: { web: true, downloadMb: 22 },
    executionMode: 'reactive',
    icon: 'Smartphone',
  },
  {
    id: 'backbone.whisper.tiny',
    name: 'Whisper tiny (encoder)',
    category: 'backbone',
    description: 'Whisper tiny audio encoder. Produces per-frame features.',
    inputs: [{ name: 'audio', type: 'audio', array: true }],
    outputs: [{ name: 'features', type: 'features', array: true }],
    params: [],
    capability: { web: true, downloadMb: 75 },
    executionMode: 'reactive',
    icon: 'AudioWaveform',
  },
  {
    id: 'backbone.mediapipe.pose',
    name: 'MediaPipe Pose',
    category: 'backbone',
    description: 'Landmark-based human pose backbone. 132-d features (33 joints × 4).',
    inputs: [{ name: 'images', type: 'image', array: true }],
    outputs: [{ name: 'features', type: 'features', array: true }],
    params: [{ name: 'modelComplexity', kind: 'enum', options: ['lite', 'full'], default: 'lite' }],
    capability: { web: true, downloadMb: 12 },
    executionMode: 'reactive',
    icon: 'PersonStanding',
  },
  {
    id: 'backbone.clip.vit-l-14',
    name: 'CLIP ViT-L/14 (image)',
    category: 'backbone',
    description: 'Larger CLIP image encoder. 768-d features, higher accuracy.',
    inputs: [{ name: 'images', type: 'image', array: true }],
    outputs: [{ name: 'features', type: 'features', array: true }],
    params: [{ name: 'normalize', kind: 'boolean', default: true }],
    capability: { localOnly: true, minVramGb: 6, downloadMb: 1700 },
    executionMode: 'reactive',
    icon: 'Brain',
  },

  // ──────────────────────────── HEAD ────────────────────────────
  {
    id: 'head.linear',
    name: 'Linear Classifier',
    category: 'head',
    description: 'Trainable linear head. Soft-max over classes.',
    inputs: [
      { name: 'features', type: 'features', array: true },
      { name: 'labels', type: 'labels', array: true, optional: true },
    ],
    outputs: [{ name: 'model', type: 'model' }],
    params: [
      { name: 'epochs', kind: 'number', default: 50, min: 1, max: 500 },
      { name: 'learningRate', kind: 'number', default: 0.001, min: 1e-5, max: 1, step: 1e-4 },
      { name: 'batchSize', kind: 'number', default: 16, min: 1, max: 256 },
    ],
    capability: { web: true },
    executionMode: 'triggered',
    icon: 'Calculator',
  },
  {
    id: 'head.knn',
    name: 'k-NN',
    category: 'head',
    description: 'Non-parametric nearest-neighbor classifier. No training required.',
    inputs: [
      { name: 'features', type: 'features', array: true },
      { name: 'labels', type: 'labels', array: true },
    ],
    outputs: [{ name: 'model', type: 'model' }],
    params: [
      { name: 'k', kind: 'number', default: 5, min: 1, max: 50 },
      { name: 'metric', kind: 'enum', options: ['cosine', 'euclidean'], default: 'cosine' },
    ],
    capability: { web: true },
    executionMode: 'triggered',
    icon: 'Network',
  },
  {
    id: 'head.zeroshot-clip',
    name: 'Zero-Shot CLIP',
    category: 'head',
    description: 'Text-prompted classifier. No labeled data needed — describe classes in words.',
    inputs: [
      { name: 'imageFeatures', type: 'features', array: true },
      { name: 'textFeatures', type: 'features', array: true },
    ],
    outputs: [{ name: 'model', type: 'model' }],
    params: [
      { name: 'temperature', kind: 'number', default: 100, min: 1, max: 1000 },
    ],
    capability: { web: true },
    executionMode: 'reactive',
    icon: 'Sparkles',
  },

  // ──────────────────────────── EVAL ────────────────────────────
  {
    id: 'eval.accuracy',
    name: 'Accuracy',
    category: 'eval',
    description: 'Top-1 accuracy on a held-out split.',
    inputs: [
      { name: 'model', type: 'model' },
      { name: 'features', type: 'features', array: true },
      { name: 'labels', type: 'labels', array: true },
    ],
    outputs: [{ name: 'metrics', type: 'metrics' }],
    params: [
      { name: 'split', kind: 'enum', options: ['val', 'test'], default: 'val' },
    ],
    capability: { web: true },
    executionMode: 'triggered',
    icon: 'Target',
  },
  {
    id: 'eval.confusion',
    name: 'Confusion Matrix',
    category: 'eval',
    description: 'Per-class confusion matrix.',
    inputs: [
      { name: 'model', type: 'model' },
      { name: 'features', type: 'features', array: true },
      { name: 'labels', type: 'labels', array: true },
    ],
    outputs: [{ name: 'metrics', type: 'metrics' }],
    params: [],
    capability: { web: true },
    executionMode: 'triggered',
    icon: 'Grid3x3',
  },

  // ──────────────────────────── DEPLOY ────────────────────────────
  {
    id: 'deploy.preview',
    name: 'Live Preview',
    category: 'deploy',
    description: 'Classify a live webcam / mic stream and show class probabilities.',
    inputs: [
      { name: 'model', type: 'model' },
      { name: 'features', type: 'features', array: true },
    ],
    outputs: [],
    params: [
      { name: 'showBars', kind: 'boolean', default: true },
      { name: 'topK', kind: 'number', default: 3, min: 1, max: 10 },
    ],
    capability: { web: true },
    executionMode: 'reactive',
    icon: 'Monitor',
  },
  {
    id: 'deploy.export-onnx',
    name: 'Export ONNX',
    category: 'deploy',
    description: 'Serialize the trained pipeline to ONNX.',
    inputs: [{ name: 'model', type: 'model' }],
    outputs: [],
    params: [
      { name: 'opsetVersion', kind: 'number', default: 17, min: 11, max: 20 },
      { name: 'filename', kind: 'string', default: 'wireml-model.onnx' },
    ],
    capability: { web: true },
    executionMode: 'triggered',
    icon: 'Download',
  },
  {
    id: 'deploy.share-url',
    name: 'Shareable URL',
    category: 'deploy',
    description: 'Encode the trained classifier into a URL for sharing.',
    inputs: [{ name: 'model', type: 'model' }],
    outputs: [],
    params: [],
    capability: { web: true },
    executionMode: 'triggered',
    icon: 'Link2',
  },
] as const;

/** Look up a schema by id. Throws if the id is unknown. */
export function getSchema(id: string): NodeSchema {
  const schema = NODE_SCHEMAS.find((s) => s.id === id);
  if (!schema) throw new Error(`Unknown node schema: ${id}`);
  return schema;
}

export function getSchemas(ids: readonly string[]): NodeSchema[] {
  return ids.map(getSchema);
}

File: wireml/registry.py
"""Node catalog. Every stage the TUI can execute comes from here."""
from __future__ import annotations

from wireml.schema import Capability, NodeSchema, ParamSpec, Port

NODE_SCHEMAS: tuple[NodeSchema, ...] = (
    # ───────── DATA ─────────
    NodeSchema(
        id="data.upload",
        name="Upload images",
        category="data",
        description="Folder-per-class dataset. One subfolder per class name.",
        outputs=(
            Port("images", "image", array=True),
            Port("labels", "labels", array=True),
            Port("sample_ids", "sample_ids", array=True),
        ),
        params=(
            ParamSpec(
                "folder",
                "file",
                default="",
                description="Path to a folder containing one subfolder per class",
            ),
        ),
        execution_mode="triggered",
    ),
    NodeSchema(
        id="data.synthetic",
        name="Synthetic demo data",
        category="data",
        description="Built-in toy features + labels for smoke-testing the pipeline.",
        outputs=(
            Port("features", "features", array=True),
            Port("labels", "labels", array=True),
            Port("sample_ids", "sample_ids", array=True),
        ),
        params=(
            ParamSpec("num_per_class", "number", default=40, min=4, max=500),
            ParamSpec("num_classes", "number", default=3, min=2, max=10),
            ParamSpec("feature_dim", "number", default=16, min=2, max=512),
        ),
        execution_mode="triggered",
    ),
    # ───────── BACKBONE ─────────
    NodeSchema(
        id="backbone.clip.vit-b-32",
        name="CLIP ViT-B/32",
        category="backbone",
        description="OpenAI CLIP base. 512-dim image features.",
        inputs=(Port("images", "image", array=True),),
        outputs=(Port("features", "features", array=True),),
        params=(ParamSpec("normalize", "boolean", default=True),),
        capability=Capability(download_mb=335),
        execution_mode="triggered",
    ),
    NodeSchema(
        id="backbone.clip.vit-l-14",
        name="CLIP ViT-L/14",
        category="backbone",
        description="Larger CLIP. 768-dim features, higher accuracy.",
        inputs=(Port("images", "image", array=True),),
        outputs=(Port("features", "features", array=True),),
        params=(ParamSpec("normalize", "boolean", default=True),),
        capability=Capability(min_vram_gb=6, download_mb=1700),
        execution_mode="triggered",
    ),
    NodeSchema(
        id="backbone.dinov2.small",
        name="DINOv2 ViT-S/14",
        category="backbone",
        description="Meta DINOv2. 384-dim self-supervised image features.",
        inputs=(Port("images", "image", array=True),),
        outputs=(Port("features", "features", array=True),),
        capability=Capability(download_mb=88),
        execution_mode="triggered",
    ),
    NodeSchema(
        id="backbone.identity",
        name="Identity (pass-through)",
        category="backbone",
        description="No feature extraction — use when upstream already provides features.",
        inputs=(Port("features", "features", array=True),),
        outputs=(Port("features", "features", array=True),),
        execution_mode="reactive",
    ),
    # ───────── HEAD ─────────
    NodeSchema(
        id="head.linear",
        name="Linear classifier",
        category="head",
        description="Trainable linear softmax. Fast, works with modest data.",
        inputs=(
            Port("features", "features", array=True),
            Port("labels", "labels", array=True),
            Port("sample_ids", "sample_ids", optional=True, array=True),
        ),
        outputs=(
            Port("model", "model"),
            Port("eval_features", "features", array=True),
            Port("eval_labels", "labels", array=True),
            Port("split", "split"),
        ),
        params=(
            ParamSpec("epochs", "number", default=50, min=1, max=500),
            ParamSpec("learning_rate", "number", default=0.01, min=1e-5, max=1, step=1e-4),
            ParamSpec("holdout_fraction", "number", default=0.2, min=0.05, max=0.5, step=0.05),
            ParamSpec("split_seed", "number", default=42, min=0),
        ),
    ),
    NodeSchema(
        id="head.knn",
        name="k-NN",
        category="head",
        description="Non-parametric nearest-neighbor. No training required.",
        inputs=(
            Port("features", "features", array=True),
            Port("labels", "labels", array=True),
            Port("sample_ids", "sample_ids", optional=True, array=True),
        ),
        outputs=(
            Port("model", "model"),
            Port("eval_features", "features", array=True),
            Port("eval_labels", "labels", array=True),
            Port("split", "split"),
        ),
        params=(
            ParamSpec("k", "number", default=5, min=1, max=50),
            ParamSpec("metric", "enum", default="cosine", options=("cosine", "euclidean")),
            ParamSpec("holdout_fraction", "number", default=0.2, min=0.05, max=0.5, step=0.05),
            ParamSpec("split_seed", "number", default=42, min=0),
        ),
    ),
    # ───────── EVAL ─────────
    NodeSchema(
        id="eval.accuracy",
        name="Accuracy",
        category="eval",
        description="Top-1 accuracy on a held-out split.",
        inputs=(
            Port("model", "model"),
            Port("eval_features", "features", array=True),
            Port("eval_labels", "labels", array=True),
        ),
        outputs=(Port("metrics", "metrics"),),
    ),
    NodeSchema(
        id="eval.confusion",
        name="Confusion matrix",
        category="eval",
        description="Per-class confusion matrix.",
        inputs=(
            Port("model", "model"),
            Port("eval_features", "features", array=True),
            Port("eval_labels", "labels", array=True),
        ),
        outputs=(Port("metrics", "metrics"),),
    ),
    # ───────── DEPLOY ─────────
    NodeSchema(
        id="deploy.export-onnx",
        name="Export ONNX",
        category="deploy",
        description="Serialize a trained linear head to ONNX. Needs `wireml[deploy]`.",
        inputs=(Port("model", "model"),),
        outputs=(Port("export", "export"),),
        params=(
            ParamSpec("filename", "string", default="wireml-model.onnx"),
            ParamSpec("opset", "number", default=17, min=11, max=20),
        ),
        capability=Capability(local_only=True),
    ),
)


_BY_ID: dict[str, NodeSchema] = {s.id: s for s in NODE_SCHEMAS}


def get_schema(schema_id: str) -> NodeSchema | None:
    """Look up a node schema by id. Returns None for unknown ids (per audit H5)."""
    return _BY_ID.get(schema_id)


def require_schema(schema_id: str) -> NodeSchema:
    """Like get_schema but raises KeyError. Use only where unknown-schema is a bug."""
    schema = _BY_ID.get(schema_id)
    if schema is None:
        raise KeyError(f"Unknown node schema: {schema_id}")
    return schema
